modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/TypeUtils.java

```diff
@@ -236,20 +237,36 @@ public static RelDataType getResultType(IgniteTypeFactory typeFactory, RelOptSch
      * @param origin Column origin.
      * @return Result type.
      */
-    private static Type getResultClass(IgniteTypeFactory typeFactory, RelOptSchema schema, RelDataType type,
-        @Nullable List<String> origin) {
-        if (F.isEmpty(origin))
-            return typeFactory.getResultClass(type);
+    private static Type getResultClass(
+        IgniteTypeFactory typeFactory,
+        RelOptSchema schema,
+        RelDataType type,
+        @Nullable List<String> origin
+    ) {
+        int cnt = 0; // Counter to protect from infinite recursion.
+
+        while (true) {
+            if (F.isEmpty(origin) || cnt++ >= 100)
+                return typeFactory.getResultClass(type);
```

Comment:
An exception may be like "Unsupported views depth"? Should we limit this threshold somewhere at a view creation? Like MAX_VIEW_DEPTH = 10?

Reply:
What happens if the threshold is reached? A wrong type? A type coercion or a cast error? Do we have a test for it? I tried MAX_VIEW_DEPTH == 1, MAX_VIEW_DEPTH == 0. All the related tests seem to work.